### ISSUE TITLE
Split reportCallIssue into more granular diagnostics for call argument validation

### ISSUE THREAD

**Issue by PizzoUnicredit** [DetachHead/basedpyright#1858]
### Description

# Split `reportCallIssue` into more granular diagnostics

Currently, many unrelated call-validation problems are reported through the generic `reportCallIssue` diagnostic. As a result, users who want to suppress generic overload-related diagnostics are forced to disable validation for many other useful call errors as well.

Examples:

```python
a = "hello"
a.replace("")
```

Current diagnostic:

```text
No overloads for "replace" match the provided arguments
```

Actual issue:

```text
Missing required argument "new"
```

---

```python
a = "hello"
a.replace("x", new="")
```

Current diagnostic:

```text
No overloads for "replace" match the provided arguments
```

Actual issue:

```text
Parameter "new" is positional-only
```

or:

```text
Keyword argument "new" is not allowed
```

---

```python
def func(a: str, b: str) -> None:
    pass

func("x")
```

Actual issue:

```text
Missing required argument "b"
```

---

I would like argument-related diagnostics to be separated from overload-resolution diagnostics.

Ideally, the following categories could be configured independently:

- Missing required arguments
- Unknown keyword arguments
- Positional-only parameter violations
- Too many arguments
- Overload mismatches

For example:

```json
{
  "reportMissingArgument": "error",
  "reportUnknownKeywordArgument": "error",
  "reportPositionalOnlyViolation": "error",
  "reportTooManyArguments": "error",
  "reportOverloadMismatch": "warning"
}
```

My use case is that I would like to suppress generic overload-mismatch diagnostics while still keeping validation for missing arguments, unknown arguments, positional-only violations, and similar mistakes.

Even if separate configuration options are not added, it would still be a significant usability improvement if obvious cases were reported using their actual root cause rather than being surfaced as a generic overload-resolution failure.

**Comment by DetachHead** [comment]
yeah unfortunately pyright has rules like `reportCallIssue` and even worse `reportGeneralTypeIssues` that cover a wide range of scenarios.

there have been other suggestions to split rules up, but the problem is that i want to maintain backwards compatibility by allowing users to configure basedpyright to behave the same as pyright (eg. by setting `typeCheckingMode` to `"basic"`). splitting rules up complicates this. i would have to mark `reportCallIssue` as [discouraged](https://docs.basedpyright.com/latest/configuration/config-files/#discouraged-settings_1) and keep it off in `"recommended"` and `"all"` modes, in favor of the new rules

**Comment by PizzoUnicredit** [comment]
Could be possible to add new rules that are a a split of ReportCallIssue, maintaning ReportCallIssue as it is?
If ReportCallIssue is true, the new rules are overriden and not used (ensuring backward compatibility), if it is off, the new rules can be activated, choosing a specific subset.

Could that work out?
